clang/tools/driver/CacheLauncherMode.cpp

```diff
@@ -0,0 +1,67 @@
+//===-- CacheLauncherMode.cpp - clang-cache driver mode -------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "CacheLauncherMode.h"
+#include "clang/Basic/DiagnosticCAS.h"
+#include "clang/Frontend/TextDiagnosticPrinter.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/Program.h"
+
+using namespace clang;
+
+static bool isSameProgram(const char *clangCachePath,
+                          const char *compilerPath) {
+  // Fast path check, see if they have the same parent path.
+  if (llvm::sys::path::parent_path(clangCachePath) ==
+      llvm::sys::path::parent_path(compilerPath))
+    return true;
+  // Check the file status IDs;
+  llvm::sys::fs::file_status CacheStat, CompilerStat;
+  if (llvm::sys::fs::status(clangCachePath, CacheStat))
+    return false;
+  if (llvm::sys::fs::status(compilerPath, CompilerStat))
+    return false;
+  return CacheStat.getUniqueID() == CompilerStat.getUniqueID();
+}
+
+Optional<int>
+clang::handleClangCacheInvocation(SmallVectorImpl<const char *> &Args,
+                                  llvm::StringSaver &Saver) {
+  assert(Args.size() >= 2);
+  const char *clangCachePath = Args.front();
+  // Drop initial '/path/to/clang-cache' program name.
+  Args.erase(Args.begin());
+  const char *compilerPath = Args.front();
+
+  if (isSameProgram(clangCachePath, compilerPath)) {
```

Comment:
I'll add a FIXME here, the high-level point is not whether it is the same binary or not but to determine whether the invoked compiler binary supports the caching options or not.